pkg/model/series.go

```diff
@@ -70,15 +103,141 @@ func (m *SeriesMerger) mergePoints(points []*typesv1.Point) int {
 		return points[i].Timestamp < points[j].Timestamp
 	})
 	var j int
-	for i := 1; i < l; i++ {
-		if points[j].Timestamp != points[i].Timestamp {
-			j++
-			points[j] = points[i]
+	for i := 0; i < l; i++ {
+		if m.aggregator.IsEmpty() {
+			m.aggregator.Add(points[i].Timestamp, points[i].Value)
 			continue
 		}
-		if m.sum {
-			points[j].Value += points[i].Value
+		if m.aggregator.GetTimestamp() != points[i].Timestamp {
+			points[j] = m.aggregator.GetAndReset()
+			j++
+			m.aggregator.Add(points[i].Timestamp, points[i].Value)
+			continue
 		}
+		m.aggregator.Add(points[i].Timestamp, points[i].Value)
+	}
+	if !m.aggregator.IsEmpty() {
+		points[j] = m.aggregator.GetAndReset()
 	}
 	return j + 1
 }
+
+type TimeSeriesAggregator interface {
```

Comment:
the name could be a bit confusing since we also have https://github.com/grafana/pyroscope/blob/9e55dd1235a700bfe8f471090ad7c6429a83db18/pkg/distributor/aggregator/aggregator.go#L11

I think their purpose is slightly different but putting that in the name is hard. Any thoughts @kolesnikovae?